packages/controllers/src/snaps/SnapController.ts

```diff
@@ -1034,7 +1039,28 @@ export class SnapController extends BaseController<
       }
 
       this._recordSnapRpcRequest(snapName);
-      return handler(origin, request);
+
+      // handle max request time
+      let resolve: (...args: unknown[]) => void;
+      let reject: (...args: unknown[]) => void;
+
+      const timeoutPromise = new Promise((res, rej) => {
+        resolve = res;
+        reject = rej;
+      });
+
+      const timeout = setTimeout(() => {
+        this._stopSnap(snapName);
+        reject(new Error('request timed out'));
+      }, this._maxRequestTime);
+
+      return Promise.race([
+        handler(origin, request).then((result) => {
+          clearTimeout(timeout);
+          resolve(result);
+        }),
+        timeoutPromise,
+      ]);
```

Comment:
I forget whether `.then` handlers execute synchronously when a promise is resolved, but we can avoid using `.then` entirely:

```suggestion
      // This will either get the result or reject due to the timeout.
      const result = await Promise.race([
        handler(origin, request),
        timeoutPromise,
      ]);

      clearTimeout(timeout);
      return result;
```